kafka-rest/src/test/java/io/confluent/kafkarest/testing/fake/FakeKafkaProducer.java

```diff
@@ -0,0 +1,165 @@
+    throw new UnsupportedOperationException("sendOffsetsToTransaction is not supported");
+  }
+
+  @Override
+  public void sendOffsetsToTransaction(
+      Map<TopicPartition, OffsetAndMetadata> offsets, ConsumerGroupMetadata groupMetadata) {
+    throw new UnsupportedOperationException("sendOffsetsToTransaction is not supported");
+  }
+
+  @Override
+  public void commitTransaction() {
+    throw new UnsupportedOperationException("commitTransaction is not supported");
+  }
+
+  @Override
+  public void abortTransaction() throws ProducerFencedException {
+    throw new UnsupportedOperationException("abortTransaction is not supported");
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+    return send(record, (metadata, exception) -> {});
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+    Optional<ByteString> key =
+        Optional.ofNullable(record.key())
+            .map(k -> ByteString.copyFrom(keySerializer.serialize(record.topic(), k)));
+    Optional<ByteString> value =
+        Optional.ofNullable(record.value())
+            .map(v -> ByteString.copyFrom(valueSerializer.serialize(record.topic(), v)));
+    int partition;
+    if (record.partition() != null) {
+      partition = record.partition();
+    } else {
+      partition =
+          partitioner.partition(
+              record.topic(),
+              record.key(),
+              key.map(ByteString::toByteArray).orElse(null),
+              record.value(),
+              value.map(ByteString::toByteArray).orElse(null),
+              cluster.toClusterMetadata());
+    }
+    List<Header> headers =
+        Stream.of(record.headers().toArray())
+            .map(
+                header ->
+                    new Header(
+                        header.key(),
+                        Optional.ofNullable(header.value()).map(ByteString::copyFrom)))
+            .collect(toImmutableList());
+    Instant timestamp =
+        Optional.ofNullable(record.timestamp()).map(Instant::ofEpochMilli).orElse(Instant.now());
+    RecordMetadata metadata;
+    Exception exception;
+    try {
+      metadata = cluster.produce(record.topic(), partition, headers, key, value, timestamp);
+      exception = null;
+    } catch (RuntimeException e) {
+      metadata =
+          new RecordMetadata(new TopicPartition(record.topic(), partition), -1, -1, -1, -1, -1);
+      exception = e;
+    }
+    callback.onCompletion(metadata, exception);
+    if (exception == null) {
+      return KafkaFuture.completedFuture(metadata);
+    } else {
+      return KafkaFutures.failedFuture(exception);
+    }
+  }
+
+  @Override
+  public void flush() {}
+
+  @Override
+  public List<PartitionInfo> partitionsFor(String topic) {
+    return cluster.getTopic(topic).getPartitions().stream()
+        .map(partition -> partition.toPartitionInfo(cluster))
+        .collect(toImmutableList());
```

Comment:
super-nit: Does Google Java Format like that? I find it much more unreadable than the each-link-from-the-chain-on-its-own-line style.